File: forward_models/python_models/inverse_stoc_var_hyper_priors.py
#########commercial modules
import numpy as np
import scipy.stats

##############in-house modules
import inverse_priors as ip
import inverse_stoc_hyper_priors as ishp

######one-dimensional prior distributions

class inverse_stoc_var_hyper_prior(ishp.inverse_stoc_hyper_prior):
	def __init__(self, hyperprior_types, var_prior_types, prior_types, hyperprior_params, var_prior_params, prior_hyperparams, hyper_dependence_lengths, var_dependence_lengths, dependence_lengths, param_hyperprior_types, var_param_prior_types, param_prior_types, n_stoc, n_stoc_var, n_dims):
		"""
		likelihood variances are also stochastic. n.b. their treatment is essentially same as nn parameters,
		but are treated separately for clarity.
		granularity of likelihood variances is either one var for all outputs or one var for each output.
		"""
		ishp.inverse_stoc_hyper_prior.__init__(self, hyperprior_types, prior_types, hyperprior_params, prior_hyperparams, hyper_dependence_lengths, dependence_lengths, param_hyperprior_types, param_prior_types, n_stoc, n_dims)
		assert (len(var_prior_types) == len(var_prior_params)), "length of var_prior_types and var_prior_params should be same"
		assert (len(var_dependence_lengths) == len(var_param_prior_types)), "var_dependence_lengths and var_param_prior_types should be the same length"		
		assert np.all(np.array(var_dependence_lengths, dtype = np.int64) == 1), "var_dependence_lengths must be [1] or [1]*n_stoc_var" 
		if len(var_dependence_lengths) != 1:
			assert (np.sum(var_dependence_lengths) == n_stoc_var), "in case of dependent var parameters, sum of var_dependence_lengths should equal n_stoc_var"
		self.var_prior_types = var_prior_types
		self.var_prior_params = var_prior_params
		self.n_stoc_var = n_stoc_var
		self.var_prior_ppfs = []
		self.params = np.zeros(n_stoc + n_stoc_var + n_dims)
		self.get_var_prior_ppf_objs()
		if (len(var_param_prior_types) == 1) and (len(var_dependence_lengths) == 1):
			self.var_prior_call = self.var_prior_call_ind_same
		else:
			self.var_param_prior_types = var_param_prior_types
			self.var_dependence_lengths = var_dependence_lengths
			self.var_prior_call = self.var_prior_call_by_dependence_lengths

	def get_var_prior_ppf_objs(self):
		"""
		same as get_ppf_objs in inverse_prior class, but
		for lhood variance param priors
		"""
		for i, p_type in enumerate(self.var_prior_types):
			var_prior_hyperparam1 = self.var_prior_params[i][0]
			var_prior_hyperparam2 = self.var_prior_params[i][1]
			if p_type == 0:
				self.var_prior_ppfs.append(ip.uniform_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 1:
				self.var_prior_ppfs.append(ip.pos_log_uniform_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 2:
				self.var_prior_ppfs.append(ip.neg_log_uniform_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 3:
				self.var_prior_ppfs.append(ip.log_uniform_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 4:
				self.var_prior_ppfs.append(ip.gaussian_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 5:
				self.var_prior_ppfs.append(ip.laplace_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 6:
				self.var_prior_ppfs.append(ip.cauchy_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 7:
				self.var_prior_ppfs.append(ip.delta_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 8:
				self.var_prior_ppfs.append(ip.gamma_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 9:
				self.var_prior_ppfs.append(ip.sqrt_recip_gamma_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 10:
				self.var_prior_ppfs.append(ip.recip_gamma_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 11:
				self.var_prior_ppfs.append(ip.sorted_uniform_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 12:
				self.var_prior_ppfs.append(ip.sorted_pos_log_uniform_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 13:
				self.var_prior_ppfs.append(ip.sorted_neg_log_uniform_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 14:
				self.var_prior_ppfs.append(ip.sorted_log_uniform_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 15:
				self.var_prior_ppfs.append(ip.sorted_gaussian_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 16:
				self.var_prior_ppfs.append(ip.sorted_laplace_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 17:
				self.var_prior_ppfs.append(ip.sorted_cauchy_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 18:
				self.var_prior_ppfs.append(ip.sorted_delta_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 19:
				self.var_prior_ppfs.append(ip.sorted_gamma_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 20:
				self.var_prior_ppfs.append(ip.sorted_sqrt_rec_gam_prior(var_prior_hyperparam1, var_prior_hyperparam2))
			elif p_type == 21:
				self.var_prior_ppfs.append(ip.sorted_rec_gam_prior(var_prior_hyperparam1, var_prior_hyperparam2))
	
	def __call__(self, hypercube):
		self.hyperprior_call(hypercube[:self.n_stoc])
		# print "det hyperparams"
		# print self.hyperparams
		# print "stoc hyperparams"
		# print self.stoc_hyperparams
		self.var_prior_call(hypercube[self.n_stoc: self.n_stoc + self.n_stoc_var])
		return self.prior_call(hypercube[self.n_stoc + self.n_stoc_var:])

	def var_prior_call_ind_same(self, hypercube):
		# print "hypercube"
		# print hypercube
		# print 
		self.params[self.n_stoc: self.n_stoc + self.n_stoc_var] = self.var_prior_ppfs[0](hypercube)
		# print "var params"
		# print self.params[self.n_stoc: self.n_stoc + self.n_stoc_var]
		return self.params

	def var_prior_call_by_dependence_lengths(self, hypercube):
		start_ind = 0
		for i, dependence_length in enumerate(self.var_dependence_lengths):
			self.params[start_ind + self.n_stoc: start_ind + self.n_stoc + dependence_length] = self.var_prior_ppfs[self.var_param_prior_types[i]](hypercube[start_ind: start_ind + dependence_length])
			start_ind += dependence_length
		# print "var params"
		# print self.params[self.n_stoc: self.n_stoc + self.n_stoc_var]
		return self.params

	def prior_call_ind_same(self, hypercube):
		"""
		adapted from inverse_stoc_hyper_priors class to account for stoc vars in indexing
		"""
		self.params[self.n_stoc + self.n_stoc_var:] = self.prior_ppfs[0](hypercube, self.hyperparams, self.stoc_hyperparams)
		# print "params"
		# print self.params[self.n_stoc + self.n_stoc_var:]
		# print "total params"
		# print self.params
		return self.params

	def prior_call_by_par_type(self, hypercube):
		"""
		adapted from inverse_stoc_hyper_priors class to account for stoc vars in indexing
		"""
		pass

	def prior_call_by_dependence_lengths(self, hypercube):
		"""
		adapted from inverse_stoc_hyper_priors class to account for stoc vars in indexing
		"""
		start_ind = 0
		for i, dependence_length in enumerate(self.dependence_lengths):
			self.params[start_ind + self.n_stoc + self.n_stoc_var: start_ind + self.n_stoc + self.n_stoc_var + dependence_length] = self.prior_ppfs[self.param_prior_types[i]](hypercube[start_ind: start_ind + dependence_length], self.hyperparams[start_ind: start_ind + dependence_length], self.stoc_hyperparams[start_ind: start_ind + dependence_length])
			start_ind += dependence_length
		# print "params"
		# print self.params[self.n_stoc + self.n_stoc_var:]
		# print "total params"
		# print self.params
		return self.params

File: forward_models/cpp_models/src/inverse_stoc_var_hyper_priors.cpp
/* external codebases */
#include <Eigen/Dense>
#include <iostream>

/* in-house code */
#include "inverse_priors.hpp"
#include "inverse_stoc_hyper_priors.hpp"
#include "inverse_stoc_var_hyper_priors.hpp"

using namespace std::placeholders;
svh_inverse_prior::svh_inverse_prior(std::vector<uint> hyperprior_types_, std::vector<uint> var_prior_types_, std::vector<uint> prior_types_, std::vector<double> hyperprior_params_, std::vector<double> var_prior_params_, std::vector<double> prior_hyperparams_, std::vector<uint> hyper_dependence_lengths_, std::vector<uint> var_dependence_lengths_, std::vector<uint> dependence_lengths_, std::vector<uint> param_hyperprior_types_, std::vector<uint> var_param_prior_types_, std::vector<uint> param_prior_types_, uint n_stoc_, uint n_stoc_var_, uint n_dims_) :
	sh_inverse_prior::sh_inverse_prior(hyperprior_types_, prior_types_, hyperprior_params_, prior_hyperparams_, hyper_dependence_lengths_, dependence_lengths_, param_hyperprior_types_, param_prior_types_, n_stoc_, n_dims_),
	var_prior_types(var_prior_types_),
	var_prior_params(var_prior_params_),
	var_dependence_lengths(var_dependence_lengths_),
	var_param_prior_types(var_param_prior_types_),
	n_stoc_var(n_stoc_var_),
	ppf_var_ptr_v(get_var_ppf_ptr_vec()) {
}

svh_inverse_prior::~svh_inverse_prior() {
	// these aren't needed here, as they are called automatically in sh destructor
	// for (std::vector<sh_base_prior *>::iterator it = ppf_ptr_v.begin(); it < ppf_ptr_v.end(); it++) {
    	// delete *it;
    // }
    // for (std::vector<base_prior *>::iterator it = ppf_hp_ptr_v.begin(); it < ppf_hp_ptr_v.end(); it++) {
    	// delete *it;
    // }
    for (std::vector<base_prior *>::iterator it = ppf_var_ptr_v.begin(); it < ppf_var_ptr_v.end(); it++) {
    	delete *it;
    }
}

std::vector<base_prior *> svh_inverse_prior::get_var_ppf_ptr_vec() {
	std::vector<base_prior *> ppf_var_ptr_vec;
	ppf_var_ptr_vec.reserve(var_prior_types.size());
	double hyperparam1;
	double hyperparam2;
	for (uint i = 0; i < var_prior_types.size(); ++i) {
		hyperparam1 = var_prior_params.at(2 * i);
		hyperparam2 = var_prior_params.at(2 * i + 1);
		if (var_prior_types.at(i) == 0) {
			ppf_var_ptr_vec.push_back(new uniform_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 1) {
			ppf_var_ptr_vec.push_back(new pos_log_uniform_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 2) {
			ppf_var_ptr_vec.push_back(new neg_log_uniform_prior(hyperparam1, hyperparam2));			
		}
		else if (var_prior_types.at(i) == 3) {
			ppf_var_ptr_vec.push_back(new log_uniform_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 4) {
			ppf_var_ptr_vec.push_back(new gaussian_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 5) {
			ppf_var_ptr_vec.push_back(new laplace_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 6) {
			ppf_var_ptr_vec.push_back(new cauchy_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 7) {
			ppf_var_ptr_vec.push_back(new delta_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 8) {
			ppf_var_ptr_vec.push_back(new gamma_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 9) {
			ppf_var_ptr_vec.push_back(new sqrt_recip_gamma_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 10) {
			ppf_var_ptr_vec.push_back(new recip_gamma_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 11) {
			ppf_var_ptr_vec.push_back(new sorted_uniform_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 12) {
			ppf_var_ptr_vec.push_back(new sorted_pos_log_uniform_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 13) {
			ppf_var_ptr_vec.push_back(new sorted_neg_log_uniform_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 14) {
			ppf_var_ptr_vec.push_back(new sorted_log_uniform_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 15) {
			ppf_var_ptr_vec.push_back(new sorted_gaussian_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 16) {
			ppf_var_ptr_vec.push_back(new sorted_laplace_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 17) {
			ppf_var_ptr_vec.push_back(new sorted_cauchy_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 18) {
			ppf_var_ptr_vec.push_back(new sorted_delta_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 19) {
			ppf_var_ptr_vec.push_back(new sorted_gamma_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 20) {
			ppf_var_ptr_vec.push_back(new sorted_sqrt_rec_gam_prior(hyperparam1, hyperparam2));
		}
		else if (var_prior_types.at(i) == 21) {
			ppf_var_ptr_vec.push_back(new sorted_rec_gam_prior(hyperparam1, hyperparam2));
		}
	}
	return ppf_var_ptr_vec;
}

void svh_inverse_prior::var_prior_call_by_dependence_lengths(Eigen::Ref<Eigen::VectorXd> cube_m, Eigen::Ref<Eigen::VectorXd> theta_m) {
	uint start_ind = 0;
	uint dependence_length;
	for (uint i = 0; i < var_dependence_lengths.size(); ++i) { 
		dependence_length = var_dependence_lengths.at(i);
		(ppf_var_ptr_v.at(var_param_prior_types.at(i)))->operator()(cube_m.segment(start_ind, dependence_length), theta_m.segment(start_ind, dependence_length));
		start_ind += dependence_length;
	}
}

void svh_inverse_prior::var_prior_call_ind_same(Eigen::Ref<Eigen::VectorXd> cube_m, Eigen::Ref<Eigen::VectorXd> theta_m) {
	(ppf_var_ptr_v.at(0))->operator()(cube_m, theta_m);
}

void svh_inverse_prior::operator()(Eigen::Ref<Eigen::VectorXd> cube_m, Eigen::Ref<Eigen::VectorXd> theta_m) {
	// std::cout << "cube" << std::endl;
	// std::cout << cube_m << std::endl;
	// std::cout << "det hyperparams" << std::endl;
	// std::cout << hyperparams_m << std::endl;
	std::function<void(Eigen::Ref<Eigen::VectorXd>, Eigen::Ref<Eigen::VectorXd>)> hyperprior_call_ptr;
	if (hyper_dependence_lengths.size() == 1) {
		hyperprior_call_ptr = std::bind(&sh_inverse_prior::hyperprior_call_ind, this, _1, _2);
	}
	else {
		hyperprior_call_ptr = std::bind(&sh_inverse_prior::hyperprior_call_by_hyper_dependence_lengths, this, _1, _2);
	}
	hyperprior_call_ptr(cube_m.segment(0, n_stoc), theta_m.segment(0, n_stoc));
	// std::cout << "stoc hyperparams" << std::endl;
	// std::cout << stoc_hyperparams_m << std::endl;
	std::function<void(Eigen::Ref<Eigen::VectorXd>, Eigen::Ref<Eigen::VectorXd>)> var_prior_call_ptr;
	if (var_dependence_lengths.size() == 1) {
		var_prior_call_ptr = std::bind(&svh_inverse_prior::var_prior_call_ind_same, this, _1, _2);
	}
	else {
		var_prior_call_ptr = std::bind(&svh_inverse_prior::var_prior_call_by_dependence_lengths, this, _1, _2);
	}
	var_prior_call_ptr(cube_m.segment(n_stoc, n_stoc_var), theta_m.segment(n_stoc, n_stoc_var));
	// std::cout << "var params" << std::endl;
	// std::cout << theta_m.segment(n_stoc, n_stoc_var) << std::endl;	
	std::function<void(Eigen::Ref<Eigen::VectorXd>, Eigen::Ref<Eigen::VectorXd>)> prior_call_ptr;
	if (dependence_lengths.size() == 1) {
		prior_call_ptr = std::bind(&sh_inverse_prior::prior_call_ind_same, this, _1, _2);
	}
	else {
		prior_call_ptr = std::bind(&sh_inverse_prior::prior_call_by_dependence_lengths, this, _1, _2);
	}
	prior_call_ptr(cube_m.segment(n_stoc + n_stoc_var, n_dims), theta_m.segment(n_stoc + n_stoc_var, n_dims));
	// std::cout << "full params" << std::endl;
	// std::cout << theta_m << std::endl;
}